Floorplan.py

```diff
@@ -37,6 +41,70 @@ def _parseDrawing(drawing):
         for (y, line) in enumerate(lines):
             for (x, c) in enumerate(line):
                 if c == '#':
-                    obstacles += [{'x': x, 'y':  y, 'width': 1, 'height': 1}] # TODO: could do line merging here to
-        return width, height, obstacles
+                    obstacles += [{'x': x, 'y':  y, 'width': 1, 'height': 1}]
+                if c == 's':
+                    (initX, initY) = (x,y)
+        return width, height, obstacles, initX, initY
+
+    def _isMapValid(self, drawing):
+        '''
+        checks if map given to floorplan is valid
+        aka. has borders and valid characters (#,., s)
+        example input - output :
+            {
+            'in': {
+                'drawing':
+                    \'''
+                    ########
+                    #...s..#
+                    ########
+                    \'''
+            },
+            'out': False
+        },
+        '''
+
+        returnVal       = True
+
+        # split drawing into matrix of characters indexed by row, column indices
+        matrixOfChars   = [line for line in drawing.splitlines() if line]
+        charsChecked    = []     # character positions in map of which surrounding frontiers have been determined
+        charsToCheck    = []     # character positions in map of which surrounding frontiers are to be determined
+
+        # if there is no starting position character 's' map is invalid
+        try:
+            startPos    = [(lidx,line.index('s')) for (lidx,line) in enumerate(matrixOfChars) if 's' in line][0]
+        except IndexError:
+            startPos    = None
+            returnVal   = False
+
+        charsToCheck   += [startPos]
+
+        # keep looping until there are no more froniters (all borders have been found)
+        while charsToCheck and startPos:
+
+            (charRow,charCol) = charsToCheck.pop(0)
+            charsChecked     += [(charRow,charCol)]
+
+            # set surrounding 8 character positions as frontiers to check next if they are not obstacles
+            neighbourChars    = [
+                (charRow+1,  charCol  ), (charRow-1,  charCol  ),
+                (charRow,    charCol-1), (charRow,    charCol+1),
+                (charRow-1,  charCol-1), (charRow+1,  charCol+1),
+                (charRow-1,  charCol+1), (charRow+1,  charCol-1)
+            ]
+
+            for nx, ny in neighbourChars:
+                
+                try:
+                    # if drawing contains characters that aren't '#,.,s' map is invalid
+                    assert matrixOfChars[nx][ny] == '#' or matrixOfChars[nx][ny] == '.' or matrixOfChars[nx][ny] == 's'
```

Comment:
do we draw the map with dots of spaces?

Reply:
I thought it's best to use dots as it's easier to differentiate between empty spaces (that may be there by mistake, wrong indentation, .. etc) and the actual free space in our maps. it also helps with debugging